File: app/api/aica_api/services/recovery.py
"""Pure recovery-phase state machine (Approach A).

Drives a RecoveryState through an option's ordered stages. No I/O, no engine
coupling — the tick engine calls advance_recovery once per tick and applies the
motion/recovery effects from the returned phase.
"""
from __future__ import annotations
from aica_api.models.run import RecoveryState, RestSpot
from aica_api.models.scenario import RecoveryOption, RecoveryStage


def current_stage(state: RecoveryState, option: RecoveryOption) -> RecoveryStage | None:
    if state.stage_index < 0 or state.stage_index >= len(option.stages):
        return None
    return option.stages[state.stage_index]


def start_recovery(option: RecoveryOption, rest_spot: RestSpot) -> RecoveryState:
    first = option.stages[0] if option.stages else None
    return RecoveryState(
        active=True,
        option_id=option.id,
        rest_spot=rest_spot,
        phase=(first.phase if first else "resuming"),
        stage_index=0,
        stage_ticks_remaining=(first.ticks or 0) if first else 0,
    )


def _enter_stage(state: RecoveryState, option: RecoveryOption, index: int) -> RecoveryState:
    # Feature 020 (Slice-2 core, review fix): reset the MOVING-stage recovery
    # accrual counters on every stage transition, so a cap_drowsiness/
    # cap_fatigue applied to a later stage's own aggregate never inherits an
    # earlier stage's accrued total.
    _reset_accrual = {
        "moving_recovery_accrued_drowsiness": 0.0,
        "moving_recovery_accrued_fatigue": 0.0,
    }
    if index >= len(option.stages):
        return state.model_copy(update={
            "phase": "resuming", "stage_index": index, "stage_ticks_remaining": 0,
            **_reset_accrual,
        })
    stage = option.stages[index]
    return state.model_copy(update={
        "phase": stage.phase, "stage_index": index, "stage_ticks_remaining": stage.ticks or 0,
        **_reset_accrual,
    })


def advance_recovery(state: RecoveryState, option: RecoveryOption, *, at_rest_spot: bool) -> RecoveryState:
    if state.phase == "resuming":
        return state.model_copy(update={"active": False, "phase": None})

    stage = current_stage(state, option)
    if stage is None:
        return state.model_copy(update={"phase": "resuming", "stage_ticks_remaining": 0})

    # MOVING wakefulness: hold until the car reaches the rest spot.
    if stage.motion == "MOVING":
        if not at_rest_spot:
            return state                      # keep driving toward the spot
        return _enter_stage(state, option, state.stage_index + 1)

    # STOPPED stage: count down its dwell ticks, then move on.
    remaining = state.stage_ticks_remaining - 1
    if remaining > 0:
        return state.model_copy(update={"stage_ticks_remaining": remaining})
    return _enter_stage(state, option, state.stage_index + 1)

File: htmlapp/frontend/src/engine/recovery.ts
/**
 * Pure recovery-phase state machine (Approach A).
 *
 * Ported from `app/api/aica_api/services/recovery.py` (behavior-of-record).
 * Drives a RecoveryState through an option's ordered stages. No I/O, no engine
 * coupling — the tick engine calls advanceRecovery once per tick and applies
 * the motion/recovery effects from the returned phase.
 *
 * Only the exported function identifiers (currentStage, startRecovery,
 * advanceRecovery) are camelCased. All object keys and phase/stage string
 * values cross the parity boundary and are preserved byte-for-byte from the
 * Python (snake_case). The Python's keyword-only `at_rest_spot` becomes the
 * options-object field `{ atRestSpot }`.
 */

import type { RecoveryOption, RecoveryStage, RecoveryStateT, RestSpot } from '../api/types'

export function currentStage(state: RecoveryStateT, option: RecoveryOption): RecoveryStage | null {
  const stages = option.stages ?? []
  if (state.stage_index < 0 || state.stage_index >= stages.length) {
    return null
  }
  return stages[state.stage_index]
}

export function startRecovery(option: RecoveryOption, restSpot: RestSpot): RecoveryStateT {
  const stages = option.stages ?? []
  const first = stages.length > 0 ? stages[0] : null
  return {
    active: true,
    option_id: option.id,
    rest_spot: restSpot,
    phase: first ? first.phase : 'resuming',
    stage_index: 0,
    stage_ticks_remaining: first ? (first.ticks ?? 0) : 0,
  }
}

function enterStage(state: RecoveryStateT, option: RecoveryOption, index: number): RecoveryStateT {
  const stages = option.stages ?? []
  if (index >= stages.length) {
    return { ...state, phase: 'resuming', stage_index: index, stage_ticks_remaining: 0 }
  }
  const stage = stages[index]
  return {
    ...state,
    phase: stage.phase,
    stage_index: index,
    stage_ticks_remaining: stage.ticks ?? 0,
  }
}

export function advanceRecovery(
  state: RecoveryStateT,
  option: RecoveryOption,
  { atRestSpot }: { atRestSpot: boolean },
): RecoveryStateT {
  if (state.phase === 'resuming') {
    return { ...state, active: false, phase: null }
  }

  const stage = currentStage(state, option)
  if (stage === null) {
    return { ...state, phase: 'resuming', stage_ticks_remaining: 0 }
  }

  // MOVING wakefulness: hold until the car reaches the rest spot.
  if (stage.motion === 'MOVING') {
    if (!atRestSpot) {
      return state // keep driving toward the spot
    }
    return enterStage(state, option, state.stage_index + 1)
  }

  // STOPPED stage: count down its dwell ticks, then move on.
  const remaining = state.stage_ticks_remaining - 1
  if (remaining > 0) {
    return { ...state, stage_ticks_remaining: remaining }
  }
  return enterStage(state, option, state.stage_index + 1)
}
